src/java/org/apache/cassandra/index/sai/plan/StorageAttachedIndexSearcher.java

```diff
@@ -0,0 +1,593 @@
+    private final QueryContext queryContext;
+    private final PrimaryKeyFactory keyFactory;
+
+    public StorageAttachedIndexSearcher(ColumnFamilyStore cfs,
+                                        TableQueryMetrics tableQueryMetrics,
+                                        ReadCommand command,
+                                        RowFilter filterOperation,
+                                        long executionQuotaMs)
+    {
+        this.command = command;
+        this.queryContext = new QueryContext(command, executionQuotaMs);
+        this.controller = new QueryController(cfs, command, filterOperation, queryContext, tableQueryMetrics);
+        this.keyFactory = PrimaryKey.factory(cfs.metadata().comparator);
+    }
+
+    @Override
+    public ReadCommand command()
+    {
+        return command;
+    }
+
+    @Override
+    public PartitionIterator filterReplicaFilteringProtection(PartitionIterator fullResponse)
+    {
+        for (RowFilter.Expression expression : controller.filterOperation())
+        {
+            AbstractAnalyzer analyzer = controller.getContext(expression).getIndexAnalyzerFactory().create();
+            try
+            {
+                if (analyzer.transformValue())
+                    return applyIndexFilter(fullResponse, analyzeFilter(), queryContext);
+            }
+            finally
+            {
+                analyzer.end();
+            }
+        }
+
+        // if no analyzer does transformation
+        return Index.Searcher.super.filterReplicaFilteringProtection(fullResponse);
+    }
+
+    @Override
+    public UnfilteredPartitionIterator search(ReadExecutionController executionController) throws RequestTimeoutException
+    {
+        return new ResultRetriever(analyze(), analyzeFilter(), controller, executionController, queryContext, keyFactory);
+    }
+
+    /**
+     * Converts expressions into filter tree for query.
+     *
+     * @return operation
+     */
+    private RangeIterator analyze()
+    {
+        return Operation.buildIterator(controller);
+    }
+
+    /**
+     * Converts expressions into filter tree (which is currently just a single AND).
+     *
+     * Filter tree allows us to do a couple of important optimizations
+     * namely, group flattening for AND operations (query rewrite), expression bounds checks,
+     * "satisfies by" checks for resulting rows with an early exit.
+     *
+     * @return root of the filter tree.
+     */
+    private FilterTree analyzeFilter()
+    {
+        return Operation.buildFilter(controller);
+    }
+
+    private static class ResultRetriever extends AbstractIterator<UnfilteredRowIterator> implements UnfilteredPartitionIterator
+    {
+        private final PrimaryKey firstPrimaryKey;
+        private final PrimaryKey lastPrimaryKey;
+        private final Iterator<DataRange> keyRanges;
+        private AbstractBounds<PartitionPosition> currentKeyRange;
+
+        private final RangeIterator operation;
```

Comment:
nit/naming: Once we've got an iterator, I'd rather name this `results`, or anything that connotes "this is an iterator of primary keys"